File: rust/codingtest/src/leet_code/linked_list/doubly_linked_list.rs
use std::rc::Rc;
use std::cell::RefCell;

//Doubly Linked List Node
#[derive(Debug)]
struct DoublyLinkedListNode<T>{
    prev : Option<Rc<RefCell<DoublyLinkedListNode<T>>>>,
    val : T,
    next : Option<Rc<RefCell<DoublyLinkedListNode<T>>>>
}
impl<T> DoublyLinkedListNode<T>{
    fn new(val : T) -> Rc<RefCell<DoublyLinkedListNode<T>>>{
        Rc::new(RefCell::new(DoublyLinkedListNode{
            prev : None,
            val,
            next : None,
        }))
    }
}
//Doubly Linked List Object
#[derive(Debug)]
struct DoublyLinkedList<T> {
    head: Option<Rc<RefCell<DoublyLinkedListNode<T>>>>,
    tail: Option<Rc<RefCell<DoublyLinkedListNode<T>>>>,
    length: usize
}

//T에 대한 Copy 허용
impl<T> DoublyLinkedList<T> where T: Copy{
    fn new() -> Self{
        DoublyLinkedList{
            head: None,
            tail: None,
            length: 0
        }
    }
    fn len(&self) -> usize{
        self.length
    }

    fn push_front(&mut self, val: T) -> Rc<RefCell<DoublyLinkedListNode<T>>>{
        let new_node = DoublyLinkedListNode::new(val);
        if let Some(old_node) = &self.head.take() {
            new_node.borrow_mut().next = Some(Rc::clone(&old_node));
            old_node.borrow_mut().prev = Some(Rc::clone(&new_node));
        }else{
            self.tail = Some(Rc::clone(&new_node));
        }
        self.head = Some(Rc::clone(&new_node));
        self.length += 1;
        new_node
    }

    fn push_back(&mut self, val: T) -> Rc<RefCell<DoublyLinkedListNode<T>>> {
        let new_tail = DoublyLinkedListNode::new(val);
        new_tail
    }
    fn pop_front(&mut self) -> Option<T>{
        match &self.head.take(){
            Some(head_node) => {
                match head_node.borrow_mut().next.as_mut(){
                    Some(next_node) =>{
                        next_node.borrow_mut().prev = None;
                        self.head = Some(Rc::clone(&next_node));
                    },
                    _ =>{
                        self.tail = None;
                    }
                }
                self.length -= 1;
                Some(head_node.borrow().val)
            },
            _ => None
        }
    }

    //back노드 연결 해제
    fn pop_back(&mut self) -> Option<T>{
        match &self.tail.take() {
            Some(tail_node) => {
                match tail_node.borrow_mut().prev.as_mut(){
                    Some(prev_node) => {
                        prev_node.borrow_mut().next = None;
                        self.tail = Some(Rc::clone(&prev_node));
                    },
                    _ => {
                        self.head = None;
                    }
                }
                self.length -= 1;
                Some(tail_node.borrow().val)
            },
            _ => None
        }
    }

    fn remove(&mut self, node: Rc<RefCell<DoublyLinkedListNode<T>>>) -> T{
        if node.borrow().prev.is_none() && node.borrow().next.is_none(){
            self.head = None;
            self.tail = None;
            self.length = 0;

            return node.borrow().val;
        }
        let mut rn = node.borrow_mut();
        match rn.prev.take() {
            Some(prev_node) => {
                match rn.next.take(){
                    Some(next_node) => {
                        prev_node.borrow_mut().next = Some(Rc::clone(&next_node));
                        next_node.borrow_mut().prev = Some(prev_node);
                    },
                    _ => {
                        prev_node.borrow_mut().next = None;
                        self.tail = Some(prev_node);
                    }
                }
            },
            _ => {
                self.head = rn.next.clone();
                self.head.as_mut().unwrap().borrow_mut().prev = None;
            }
        }
        self.length -= 1;
        rn.val
    }
}

pub fn run(){
    let mut t : DoublyLinkedList<i32> = DoublyLinkedList::new();
    t.push_front(1);
    t.push_front(2);
    println!("Doubly Linked List : {:#?} ", t.head);
}

File: typescript/base_syntax/base_ds_algorithm/linked_list/doubly_linked_list.js
"use strict";
exports.__esModule = true;
var DoublyLinkedListNode = /** @class */ (function () {
    function DoublyLinkedListNode(value) {
        this.prev = null;
        this.value = value;
        this.next = null;
    }
    return DoublyLinkedListNode;
}());
//push_front, push_back, 
//pop_front, pop_back, 
//remove, 
//display_front, display_back
var DoublyLinkedList = /** @class */ (function () {
    function DoublyLinkedList() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }
    //list의 front에 data 삽입
    DoublyLinkedList.prototype.push_front = function (value) {
        var newHead = new DoublyLinkedListNode(value);
        if (this.head === null) {
            //tail이 새로운 노드를 가리킨다.
            this.tail = newHead;
        }
        else {
            //첫번째 노드의 prev와 newHead.next연결
            this.head.prev = newHead;
            newHead.next = this.head;
        }
        this.head = newHead;
        this.length++;
        return newHead;
    };
    //list의 back에 data 삽입
    DoublyLinkedList.prototype.push_back = function (value) {
        var newTail = new DoublyLinkedListNode(value);
        if (this.tail === null) {
            //head에 새로운 노드를 가리킨다
            this.head = newTail;
        }
        else {
            //마지막 노드의 next와 newtail.prev를 연결
            this.tail.next = newTail;
            newTail.prev = this.tail;
        }
        this.tail = newTail;
        this.length++;
        return newTail;
    };
    DoublyLinkedList.prototype.pop_front = function () {
        if (this.head === null)
            return null;
        if (this.head.next === null) {
            this.tail = null;
        }
        else {
            this.head = this.head.next.prev;
            this.head.next = null;
        }
        if (this.length > 0)
            this.length--;
        return this.head;
    };
    DoublyLinkedList.prototype.pop_back = function () {
        if (this.tail === null)
            return null;
        if (this.tail.prev === null) {
            this.head = null;
        }
        else {
            this.tail.next = null;
            this.tail = this.tail.prev.next;
        }
        console.log(this);
        if (this.length > 0)
            this.length--;
        return this.tail;
    };
    DoublyLinkedList.prototype.remove = function () {
        if (this.length > 0)
            this.length--;
    };
    return DoublyLinkedList;
}());
function runDoublyLinkedList() {
    var dll = new DoublyLinkedList();
    dll.push_front(1);
    dll.push_front(2);
    dll.push_front(3);
    console.log(dll);
    dll.pop_back();
    // dll.pop_back();
    // dll.pop_back();
    //console.log(dll);
}
exports["default"] = runDoublyLinkedList;
